xarray/core/concat.py

```diff
@@ -586,9 +642,33 @@ def get_indexes(name):
                     )
                     result_vars[k] = v
             else:
-                combined_var = concat_vars(
-                    vars, dim, positions, combine_attrs=combine_attrs
-                )
+                # do not concat if:
+                # 1. variable is only present in one dataset of multiple datasets and
+                # 2. dim is not a dimension of variable
+                if (
+                    dim not in variables[0].dims
+                    and len(variables) == 1
+                    and len(datasets) > 1
+                ):
+                    combined_var = variables[0]
+                # only concat if variable is in multiple datasets
+                # or if single dataset (GH1988)
+                else:
+                    combined_var = concat_vars(
+                        vars, dim, positions, combine_attrs=combine_attrs
+                    )
+                    # reindex if variable is not present in all datasets
+                    if len(variable_index) < len(concat_index):
+                        try:
+                            fill = fill_value[name]
+                        except (TypeError, KeyError):
+                            fill = fill_value
+                        combined_var = (
+                            DataArray(data=combined_var, name=name)
+                            .assign_coords({dim: variable_index})
+                            .reindex({dim: concat_index}, fill_value=fill)
+                            .variable
```

Comment:
I think we should be able to 
1. Assign `combined_var` *without* reindexing to `result_vars`, and
2. use `xarray.core.alignment.reindex_variables` directly on `result_vars` 

to save a lot of overhead.

Reply:
Ah, great, I thought that there might be a more fitting solution (as you mentioned in the link in the other PR), but didn't find it.